src/main/java/graphql/schema/validation/TypeAndFieldRule.java

```diff
@@ -0,0 +1,274 @@
+        } else if (type instanceof GraphQLInputObjectType) {
+            validateInputObject((GraphQLInputObjectType) type, errorCollector);
+        } else if (type instanceof GraphQLScalarType) {
+            validateScalar((GraphQLScalarType) type, errorCollector);
+        }
+    }
+
+    private void validateObject(GraphQLObjectType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+
+        List<GraphQLFieldDefinition> fieldDefinitions = type.getFieldDefinitions();
+        if (fieldDefinitions == null || fieldDefinitions.size() == 0) {
+            SchemaValidationError validationError = new SchemaValidationError(SchemaValidationErrorType.ObjectTypeLackOfFieldError, String.format("\"%s\" must define one or more fields.", type.getName()));
+            errorCollector.addError(validationError);
+            return;
+        }
+
+        if (fieldDefinitions != null && fieldDefinitions.size() != 0) {
+            for (GraphQLFieldDefinition fieldDefinition : fieldDefinitions) {
+                validateFieldDefinition(type.getName(), fieldDefinition, errorCollector);
+            }
+        }
+    }
+
+    private void validateInterface(GraphQLInterfaceType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+        List<GraphQLFieldDefinition> fieldDefinitions = type.getFieldDefinitions();
+        if (fieldDefinitions == null || fieldDefinitions.size() == 0) {
+            SchemaValidationError validationError = new SchemaValidationError(SchemaValidationErrorType.InterfaceLackOfFieldError, type.getName() + " must define one or more fields.");
+            errorCollector.addError(validationError);
+        }
+
+        if (fieldDefinitions != null && fieldDefinitions.size() != 0) {
+            for (GraphQLFieldDefinition fieldDefinition : fieldDefinitions) {
+                validateFieldDefinition(type.getName(), fieldDefinition, errorCollector);
+            }
+        }
+    }
+
+    private void validateInputObject(GraphQLInputObjectType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+        if (type.getFieldDefinitions() == null || type.getFieldDefinitions().size() == 0) {
+            SchemaValidationError validationError = new SchemaValidationError(SchemaValidationErrorType.InputObjectTypeLackOfFieldError, type.getName() + " must define one or more fields.");
+            errorCollector.addError(validationError);
+        }
+
+    }
+
+    private void validateUnion(GraphQLUnionType type, SchemaValidationErrorCollector errorCollector) {
+        assertTypeName(type.getName(), errorCollector);
+
+        UnionTypeDefinition definition = type.getDefinition();
+        if (definition == null) {
+            return;
+        }
+
+        List<Type> memberTypes = definition.getMemberTypes();
```

Comment:
we should not need to check the AST definition here. Just getting the members of the type should be enough.

Reply:
`types` in `GraphQLUnionType.Builder` is a Map([code link](https://github.com/graphql-java/graphql-java/blob/10eeacc3979624964ef10ea1078ef84534d9c5cc/src/main/java/graphql/schema/GraphQLUnionType.java#L195)), which will lost the type with duplicated name. 

Maybe it will be better to move this check to `SchemaTypeChecker`.

Reply:
I will followed your suggestion, because types definied in GraphQLUnionType is list. 

Then I will add `unique member check` to `SchemaTypeChecker` in another PR.